File: src/substitution/Vignere.kt
package substitution

class Vignere(key: String) {

    private val keyUpperCase = key.toUpperCase()

    fun encipher(plainText: String): String {
        val lowerCaseText = plainText.toUpperCase()
        val inputStream = lowerCaseText.toCharArray()
        val outputStream = mutableListOf<Char>()
        inputStream.forEachIndexed { index, letter ->
            val oldIndex = letter.toInt() - 65
            val currentKey = keyUpperCase[(index % keyUpperCase.length)].toInt() - 65
            val newIndex = ((oldIndex + currentKey) % 26) + 65
            outputStream.add(newIndex.toChar())
        }
        return outputStream.joinToString("").toUpperCase()
    }

    fun decipher(cipherText: String): String {
        val lowerCaseText = cipherText.toUpperCase()
        val inputStream = lowerCaseText.toCharArray()
        val outputStream = mutableListOf<Char>()
        inputStream.forEachIndexed { index, letter ->
            val oldIndex = letter.toInt() - 65
            val currentKey = keyUpperCase[(index % keyUpperCase.length)].toInt() - 65
            val newIndex = ((oldIndex - currentKey + 26) % 26) + 65
            outputStream.add(newIndex.toChar())
        }
        return outputStream.joinToString("")
    }
}

File: C/substitution/vignere.c
#include "substitution.h"
#include "vignere.h"

void encrypt_vignere(char *plain_text, char *cipher_text, char *key) {
    int i = 0;
    while (plain_text[i] != 0) {
        int current_index = toupper(plain_text[i]) - 65;
        int current_key = toupper(key[i % strlen(key)]) - 65;
        int cipher_letter = ((current_index + current_key) % 26) + 65;
        cipher_text[i] = (char) cipher_letter;
        i++;
    }
    cipher_text[i] = 0;
}

void decrypt_vignere(char *cipher_text, char *plain_text, char *key) {
    int i = 0;
    while (cipher_text[i] != 0) {
        int current_index = toupper(cipher_text[i]) - 65;
        int current_key = toupper(key[i % strlen(key)]) - 65;
        int plain_letter = ((current_index - current_key + 26) % 26) + 65;
        plain_text[i] = (char) plain_letter;
        i++;
    }
    plain_text[i] = 0;
}